Improve Terminal Output Formatting in pytest-cov

tsvikas opened the pull request:
This PR improves the formatting of separator lines in coverage reports and enhances the robustness of terminal width handling.

#### Why?
- To conform to the visuals from pytest headings.
- To allow clear separation of this plugin output from the previous section

#### Changes:  
- Add a new section heading for the coverage tests output.
- Make all headings the same width as pytest's headings.
- Change separator character of regular headings from `-` to `_`.

webknjaz commented on src/pytest_cov/plugin.py:
```diff
@@ -11,6 +11,11 @@
 
+try:
+    from pytest import TerminalReporter  # noqa: PT013
+except ImportError:
```
Why would this be needed?

tsvikas replied:
I wasn't sure if importing from _pytest is an acceptable thing. if it is, than those lines are useless.

webknjaz replied:
The private namespace exists specifically for imports within the project but the end-users aren't supposed to import from it.

The-Compiler replied:
I'm confused how this would change anything about it being private. `pytest.TerminalReporter` is not public, so with what you have here, you have both importing private API *and* dead code that doesn't actually do anything useful.

tsvikas replied:
`TerminalReporter` was made public in [this commit](https://github.com/pytest-dev/pytest/commit/7ce3222764193feb581abfea58f8920b36e259de), so this code uses the public API for newer versions while falling back to the private package for older ones.  

The goal was to ensure consistent styling with future pytest changes.  

I can refactor it to avoid `TerminalReporter` if preferred.

webknjaz reviewed (commented):
It'd be useful to see some output examples in the PR description, including comparison. Also, such changes usually require prior discussion.

tsvikas commented:
Thanks for the feedback! I preferred to get things going so the code and output would be visible when discussing it.

I’ve added an example of the updated output to the PR description for clarity. Let me know if you have any concerns!


Output example (with a terminal width of 100)

Before
```
======================================= test session starts ========================================
platform linux -- Python 3.13.1, pytest-8.3.4, pluggy-1.5.0
benchmark: 5.0.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /tmp/py-temp
configfile: pyproject.toml
plugins: cov-6.0.0, benchmark-5.0.1, loguru-0.4.0, reverse-1.8.0, clarity-1.0.1
collected 4 items

tests/test_cli.py .F.                                                                        [ 75%]
tests/test_.py .                                                                             [100%]

============================================= FAILURES =============================================
_________________________________________ test_app_version _________________________________________

    def test_app_version() -> None:
        result = runner.invoke(app, ["--version"])
        assert result.exit_code == 0
        assert __version__ in result.stdout
>       assert False
E       assert False

result     = <Result okay>

tests/test_cli.py:22: AssertionError

---------- coverage: platform linux, python 3.13.1-final-0 -----------
Name                    Stmts   Miss  Cover
-------------------------------------------
src/ptemp/__init__.py       3      0   100%
src/ptemp/cli.py           19      0   100%
-------------------------------------------
TOTAL                      22      0   100%

===================================== short test summary info ======================================
FAILED tests/test_cli.py::test_app_version - assert False
=================================== 1 failed, 3 passed in 0.33s ====================================

```

After:
```
======================================= test session starts ========================================
platform linux -- Python 3.13.1, pytest-8.3.4, pluggy-1.5.0
benchmark: 5.0.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /tmp/py-temp
configfile: pyproject.toml
plugins: cov-6.0.0, benchmark-5.0.1, loguru-0.4.0, reverse-1.8.0, clarity-1.0.1
collected 4 items

tests/test_cli.py .F.                                                                        [ 75%]
tests/test_.py .                                                                             [100%]

============================================= FAILURES =============================================
_________________________________________ test_app_version _________________________________________

    def test_app_version() -> None:
        result = runner.invoke(app, ["--version"])
        assert result.exit_code == 0
        assert __version__ in result.stdout
>       assert False
E       assert False

result     = <Result okay>

tests/test_cli.py:22: AssertionError

========================================= coverage report ==========================================
_________________________ coverage: platform linux, python 3.13.1-final-0 __________________________
Name                    Stmts   Miss  Cover
-------------------------------------------
src/ptemp/__init__.py       3      0   100%
src/ptemp/cli.py           19      0   100%
-------------------------------------------
TOTAL                      22      0   100%

===================================== short test summary info ======================================
FAILED tests/test_cli.py::test_app_version - assert False
=================================== 1 failed, 3 passed in 0.18s ====================================

```

The-Compiler commented on src/pytest_cov/plugin.py:
```diff
@@ -333,6 +338,8 @@ def pytest_runtestloop(self, session):
             from coverage.misc import CoverageException
 
+            tr = TerminalReporter(session.config, self.cov_report)
+            tr.write_sep('=', 'coverage report')
```
Why does this need to be added to the `StringIO` at all? If you instead output the separator in `pytest_terminal_summary`, couldn't you use the `terminalreporter` there which is used to output the report, instead of having to create a new one?

The-Compiler commented:
I agree your proposed output looks nicer (and indeed it helps to have a "preview"!), just one nitpick about it: With

```
========================================= coverage report ==========================================
_________________________ coverage: platform linux, python 3.13.1-final-0 __________________________
Name                    Stmts   Miss  Cover
-------------------------------------------
src/ptemp/__init__.py       3      0   100%
src/ptemp/cli.py           19      0   100%
-------------------------------------------
TOTAL                      22      0   100%
```

do the two separators really make sense, given that there will only ever be one coverage report (I think?). I'd have expected something like:

```
========================================= coverage report ==========================================
Name                    Stmts   Miss  Cover
-------------------------------------------
src/ptemp/__init__.py       3      0   100%
src/ptemp/cli.py           19      0   100%
-------------------------------------------
TOTAL                      22      0   100%

platform linux, python 3.13.1-final-0
```

or maybe just

```
========================= coverage: platform linux, python 3.13.1-final-0 ==========================
Name                    Stmts   Miss  Cover
-------------------------------------------
src/ptemp/__init__.py       3      0   100%
src/ptemp/cli.py           19      0   100%
-------------------------------------------
TOTAL                      22      0   100%
```

tsvikas commented:
As per request, I moved the heading to `pytest_terminal_summary`.  It seems to fit there better :).

It also helped me notice two small behaviors that I changed (I can revert them if needed):  

* Renamed the heading from "coverage report" to "tests coverage" (since in some cases, it's not a full report but a warning).  
* Adjusted newlines to match pytest's style (one newline after `_` headings, no newlines before/after `=` headings).

tsvikas commented:
current output (120 columns):
```
================================================= test session starts ==================================================
platform linux -- Python 3.13.1, pytest-8.3.4, pluggy-1.5.0
benchmark: 5.0.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /tmp/py-temp
configfile: pyproject.toml
plugins: cov-6.0.0, benchmark-5.0.1, loguru-0.4.0, reverse-1.8.0, clarity-1.0.1
collected 5 items

tests/test_.py .                                                                                                 [ 20%]
tests/test_cli.py .FF.                                                                                           [100%]

======================================================= FAILURES =======================================================
___________________________________________________ test_app_version ___________________________________________________

    def test_app_version() -> None:
        result = runner.invoke(app, ["--version"])
        assert result.exit_code == 0
        assert __version__ in result.stdout
>       assert False
E       assert False

result     = <Result okay>

tests/test_cli.py:22: AssertionError
__________________________________________________ test_app_version2 ___________________________________________________

    def test_app_version2() -> None:
        result = runner.invoke(app, ["--version"])
        assert result.exit_code == 0
        assert __version__ in result.stdout
>       assert False
E       assert False

result     = <Result okay>

tests/test_cli.py:28: AssertionError
==================================================== tests coverage ====================================================
___________________________________ coverage: platform linux, python 3.13.1-final-0 ____________________________________

Name                    Stmts   Miss  Cover
-------------------------------------------
src/ptemp/__init__.py       3      0   100%
src/ptemp/cli.py           19      0   100%
-------------------------------------------
TOTAL                      22      0   100%
=============================================== short test summary info ================================================
FAILED tests/test_cli.py::test_app_version - assert False
FAILED tests/test_cli.py::test_app_version2 - assert False
============================================= 2 failed, 3 passed in 0.18s ==============================================
```

tsvikas commented:
> I agree your proposed output looks nicer (and indeed it helps to have a "preview"!), just one nitpick about it: With

Thank you :)

> 
> ```
> ========================================= coverage report ==========================================
> _________________________ coverage: platform linux, python 3.13.1-final-0 __________________________
> Name                    Stmts   Miss  Cover
> -------------------------------------------
> src/ptemp/__init__.py       3      0   100%
> src/ptemp/cli.py           19      0   100%
> -------------------------------------------
> TOTAL                      22      0   100%
> ```
> 
> do the two separators really make sense, given that there will only ever be one coverage report (I think?). I'd have expected something like:
> 
> ```
> ========================================= coverage report ==========================================
> Name                    Stmts   Miss  Cover
> -------------------------------------------
> src/ptemp/__init__.py       3      0   100%
> src/ptemp/cli.py           19      0   100%
> -------------------------------------------
> TOTAL                      22      0   100%
> 
> platform linux, python 3.13.1-final-0
> ```
> 
> or maybe just
> 
> ```
> ========================= coverage: platform linux, python 3.13.1-final-0 ==========================
> Name                    Stmts   Miss  Cover
> -------------------------------------------
> src/ptemp/__init__.py       3      0   100%
> src/ptemp/cli.py           19      0   100%
> -------------------------------------------
> TOTAL                      22      0   100%
> ```

I saw in `engine.py` that `self.sep`, the function responsible for creating `_` headings, can produce multiple variations, and, in the event of a failed worker, it generate two `_` headings.

I believe using a constant `=` heading for the whole section would be beneficial.

Modifying or rearranging the `_` lines or the rest of the output content might be beneficial, but I assumed that it is beyond the scope of this PR, and would require input from the package owners.

it might be a good idea to change how the `_` behaves

tsvikas commented:
Hey @The-Compiler , just checking in—do you have any further feedback on this? I’ve incorporated the suggested changes and would love to get this moving forward. Let me know if anything else needs adjusting!

Also, I see there's a workflow approval pending—could someone with merge access take a look?

The-Compiler commented on src/pytest_cov/plugin.py:
```diff
@@ -217,6 +217,7 @@ def __init__(self, options, pluginmanager, start=True, no_cov_should_warn=False)
         self._start_path = None
         self._disabled = False
         self.options = options
+        self.wrote_heading = False
```
Should probably be private?

```suggestion
        self._wrote_heading = False
```

tsvikas replied:
done

The-Compiler commented on src/pytest_cov/engine.py:
```diff
@@ -134,15 +135,36 @@ def get_node_desc(platform, version_info):
-            sep_extra = sep_total % 2
-            out = f'{s * sep_len} {txt} {s * (sep_len + sep_extra)}\n'
-            stream.write(out)
+            fullwidth = self.get_width()
```
You might want to add a comment somewhere around here that this is essentially copied from `_pytest._io.terminalwriter`. Personally I would even put a note there from which commit of pytest it's coming from, that would make it easier to upgrade things (without having to dig into history) if the pytest logic ever changes.

Another possibility entirely would be to import the things from pytest despite it being private. I suppose both have their maintenance cost, personally I would have a slight preference for copying the code like you did. But credit where credit is due :wink:.

(In theory there is license considerations as well. But then again, this repo uses the same license as pytest at least, and it's questionable whether "Copyright (c) 2004 Holger Krekel and others" would need to be here for those couple lines of code. I'm... surprised how pytest-cov [is apparently copyrighted](https://github.com/pytest-dev/pytest-cov/blob/master/LICENSE) by "Meme Dough", so I feel like there are bigger problems there anyways :sweat_smile:)

tsvikas replied:
good idea. done :)

The-Compiler reviewed (changes requested):
Two more small considerations. Other than that, this looks good to me, but I'll leave merging to someone who's more involved with pytest-cov than I am.

tsvikas commented:
Thanks, @The-Compiler! I've addressed all the changes. I noticed the @maintainers link isn't working. Is there a way to ping someone for merging?

The-Compiler commented:
As a general rule, please don't ping maintainers in their own projects unless it's something absolutely critical. You can expect maintainers to be subscribed to all activity in their own projects, and/or to be periodically checking things on their own schedule.

Remember that projects are often maintained by a single person in their spare time, often with various other projects on their hands as well. You won't be the only person pinging them in a typical week, or sometimes even day. It usually make things far more stressful than they would be otherwise. Sometimes people are busy with other stuff, or on holidays, or have a tight work deadline, and then open source things will have to wait for a bit.

tsvikas commented:
Hey, just a quick check-in—I've addressed all requested changes, and the failing tests seem unrelated to this PR. Let me know if anything else is needed before merging. Thanks!

webknjaz commented:
While the failing tests may be unrelated, they may still block the PR from being merged. Sometimes, the fastest way of solving speeding up the process is making another PR fixing those problems and then rebasing.

Most projects won't be force-merging PRs with red CI and would wish to keep the main branch green before anything is allowed in.

The link check likely fails because GH changed the way they inject anchors into web pages. Sphinx has a toggle to skip this for certain websites. That would be the solution. But I'd also check that the linked thing still exists.

tsvikas commented:
Hi Team,

I've rebased this PR to the latest master, and all checks are now passing successfully. The previous issues with failing tests have been resolved.

I'm happy to provide more information or adjustments if needed. Please let me know if there's anything else required.

Thanks for your support!

webknjaz commented:
You just have to wait patiently now and hope the primary maintainers get to it eventually.
